apps/src/lib/node/ledger/shell/finalize_block.rs

```diff
@@ -50,6 +60,201 @@ where
         let (height, new_epoch) =
             self.update_state(req.header, req.hash, req.byzantine_validators);
 
+        if new_epoch {
+            for id in std::mem::take(&mut self.proposal_data) {
+                let proposal_funds_key = gov_storage::get_funds_key(id);
+
+                let funds = self
+                    .storage
+                    .read(&proposal_funds_key)
+                    .map(|(bytes, _gas)| {
+                        if let Some(bytes) = bytes {
+                            match token::Amount::try_from_slice(&bytes[..]) {
+                                Ok(amount) => amount,
+                                Err(_) => token::Amount::default(),
+                            }
+                        } else {
+                            token::Amount::default()
+                        }
+                    })
+                    .map_err(|e| Error::BadProposal(e.to_string()))?;
+
+                let tally_result = compute_tally(&self.storage, id);
+                if let Ok(tally_result) = tally_result {
+                    match tally_result {
+                        (TallyResult::Passed, Some(code)) => {
+                            let tx = Tx::new(code, Some(encode(&id)));
+                            let tx_type =
+                                TxType::Decrypted(DecryptedTx::Decrypted(tx));
+                            let pending_execution_key =
+                                gov_storage::get_proposal_execution_key(id);
+                            // in order to filter if the vp has been triggered
+                            // by the protocol or by a transaction,
+                            // we set the following key in a protected storage
+                            // space as a flag. After the execution, we delete
+                            // it.
+                            self.storage
+                                .write(&pending_execution_key, "")
+                                .expect("Should be able to write to storage.");
+                            let tx_result = protocol::apply_tx(
+                                tx_type,
+                                0,
+                                &mut BlockGasMeter::default(),
+                                &mut self.write_log,
+                                &self.storage,
+                                &mut self.vp_wasm_cache,
+                                &mut self.tx_wasm_cache,
+                            );
+                            self.storage.delete(&pending_execution_key).expect(
+                                "Should be able to delete the storage.",
+                            );
+                            match tx_result {
```

Comment:
If the tx is rejected or there's an error, we have to clear out the write log (`write_log.drop_tx()`). 

We also have to call `self.gas_meter.finalize_transaction()` every time (this doesn't depend on the result), otherwise the gas used by the proposal tx would count towards the first block transaction.

Reply:
sry, nvm `self.gas_meter.finalize_transaction()` - I forgot we're using a different meter for this